src/hx711.toit

```diff
@@ -0,0 +1,178 @@
+  /// Read from channel A with a gain of 64. 
+  static CHANNEL_A_GAIN_64  ::= Hx711Input.private_ 27
+
+  /**
+  The $clock pin is used as an output pin, connected to the UART hardware.
+  The $data pin is used as an input GPIO pin.
+  */
+  constructor --clock/int=17 --data/int=16:
+    data_ = gpio.Pin.in 16
+    clock_pin := gpio.Pin.out clock
+    uart_port_ = uart.Port
+      --tx=clock_pin
+      --rx=null
+      // This common baud rate gives an 8.6us pulse for each start bit, which is
+      // in the range of 0.1us to 50us that the HX711 requires.
+      --baud_rate=115200
+      // We want the idle state to be low and the start bit to be high.
+      --invert_tx=true
+
+  // The actual pattern on the UART TX pin will be a start bit (high,
+  // 8.6us), eight data bits (low, since the TX pin is inverted) and a stop bit
+  // (low).
+  SINGLE_PULSE_ := #[0xff]
+
+  /**
+  Returns a value between -1.0 and 1.0.
+  May also return -Infinity or Infinity if the measurement is out of range.
+  For $input pass $Hx711.CHANNEL_A_GAIN_128, $Hx711.CHANNEL_B_GAIN_32, or
+    $Hx711.CHANNEL_A_GAIN_64.
+  */
+  get input/Hx711Input -> float:
+    if input_ != input:
+      // If the input setting is wrong we do a dummy read to set the input
+      // correctly.
+      data_.wait_for 0
+      input.pulses.repeat:
+        uart_port_.write SINGLE_PULSE_
+        sleep --ms=1
+      input_ = input
+
+    // Wait for a sample to be ready.
+    data_.wait_for 0
+
+    measurement := 0
+
+    24.repeat: 
+      uart_port_.write SINGLE_PULSE_
+      sleep --ms=1
+      measurement <<= 1
+      bit := data_.get
+      measurement |= bit
+
+    // The number of pulses (25 to 27) determines the input setting of the next
+    // reading.
+    (input_.pulses - 24).repeat:
+      uart_port_.write SINGLE_PULSE_
+      sleep --ms=1
+
+    if measurement == 0: return 0.0
+
+    if measurement < 0x80_0000:
+      if measurement == 0x7f_ffff:
+        return float.INFINITY
+      else:
+        return measurement.to_float / 0x80_0000
+    else:
+      if measurement == 0x80_0000:
+        return -float.INFINITY
+      else:
+        return (measurement - 0x100_0000).to_float / 0x80_0000
+
+  /**
+  Takes 10 samples, discarding infinities unless there are a lot of
+    infinities.
+  Discards outliers using the method of removing elements that are further
+    from the mean than 1.5 interquartile ranges.  Returns the mean of the
+    non-discarded samples.
+  For $input pass $Hx711.CHANNEL_A_GAIN_128, $Hx711.CHANNEL_B_GAIN_32, or
+    $Hx711.CHANNEL_A_GAIN_64.
+  */
+  average_of_10 input/Hx711Input -> num:
+    while true:
+      infinities := 0
+      samples := []
+      while samples.size < 10:
+        measurement := get input
+        if -float.INFINITY < measurement < float.INFINITY:
+          samples.add measurement
+        else:
+          infinities++
+          if infinities > 20: return measurement
+          samples = []
+      while samples.size >= 2:
```

Comment:
Feels weird that this function could run infinitely.
If the sample size goes down to 1, we restart reading 10 fresh samples. That could happen over and over again. No?
(If it can't happen, we don't need the `while` condition).